### PR TITLE
INREL-4431 Activate Coupon

### PR DESCRIPTION
**Pull request by danschalow**
## [INREL-4431](https://jira.burda.com/browse/INREL-4431) 
 - Add templates for view modes minimal, normal, maximal and extended
 - Style view modes

## How to test:
 - https://dev.freundin.de/beauty-lidstrich-ziehen-tipps

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: js/behaviors/drupal.behaviors.coupons.js [hunk 1]
```diff
@@ -38,6 +39,17 @@
           expandContainer.classList.toggle("expand");
         });
       });
+    },
+    isExpired(coupon) {
```
Comment (mgoedecke) [inline js/behaviors/drupal.behaviors.coupons.js:43]: do we need to check if expires is set?
Something like 
```
isExpired(coupon) {
   if (null === coupon.dataset.expires) {
       return false;
   }
   ...
}
```

File: templates/field/field--node--field-advertising-products--coupon.html.twig [hunk 2]
```diff
@@ -1,17 +1,17 @@
 {%- for item in items -%}
-    
+
     {# open expandable containers #}
-    {% if items | length > 2 and loop.index == 3  %}
+    {% if items | length > 2 and loop.index == 3 %}
```
Comment (mgoedecke) [inline templates/field/field--node--field-advertising-products--coupon.html.twig]: Please don't open <div> in an if statement without closing them again. This is really bad and will most definitely become a problem sometime soon…

File: templates/field/field--node--field-advertising-products--coupon.html.twig [hunk 3]
```diff
@@ -1,17 +1,17 @@
...
 
-    {# normal product rendering #}
-    {{ item.content }}
+            {# normal product rendering #}
+            {{ item.content }}
 
-    {# close expandable containers #}
-    {% if items | length > 2 and loop.index == items | length %}
+            {# close expandable containers #}
+            {% if items | length > 2 and loop.index == items | length %}
```
Comment (mgoedecke) [inline templates/field/field--node--field-advertising-products--coupon.html.twig]: Please don't close <div> in an if statement without closing them again. This is really bad and will most definitely become a problem sometime soon…

File: templates/node/coupon/node--coupon--content-reference-maximal.html.twig [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ *   in different view modes.
+ *
+ * @ingroup themeable
+ */
+#}
+
+{% extends "node--coupon--content-reference.html.twig" %}
+
+{% block coupon_teaser %}
+    {% if content.field_teaser_media | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference-maximal.html.twig:79]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference-extended.html.twig [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ *   in different view modes.
+ *
+ * @ingroup themeable
+ */
+#}
+
+{% extends "node--coupon--content-reference-maximal.html.twig" %}
+
+{% block coupon_teaser %}
+    {% if content.field_logo | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference-extended.html.twig:79]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference-maximal.html.twig [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        <div class="coupon__aspect-ratio">
+            {{ content.field_teaser_media }}
+        </div>
+    {% endif %}
+{% endblock %}
+
+{% block advertising_products %}
+    {% if content.field_advertising_products | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference-maximal.html.twig:87]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference-normal.html.twig [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+ *   in different view modes.
+ *
+ * @ingroup themeable
+ */
+#}
+
+{% extends "node--coupon--content-reference.html.twig" %}
+
+{% block coupon_logo %}
+    {% if content.field_logo | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference-normal.html.twig:79]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference-normal.html.twig [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        <div class="coupon__brand--logo__wrapper">
+            <div class="strike-through-title__wrapper">
+                <div class="coupon__logo">{{ content.field_logo }}</div>
+            </div>
+        </div>
+    {% endif %}
+{% endblock %}
+
+{% block coupon_teaser %}
+    {% if content.field_teaser_media | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference-normal.html.twig:89]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference.html.twig [hunk 9]
```diff
@@ -0,0 +1,78 @@
+{% set classes = [
+'item-content',
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference.html.twig]: use indention

File: templates/node/coupon/node--coupon--content-reference.html.twig [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+{{ attach_library('infinite/coupons') }}
+
+<div{{ attributes.addclass(classes) }}>
+    <div class="coupon" data-expires="{{ end_date }}">
+        {% block coupon_logo %}{% endblock %}
+        <h3 class="coupon__title">{{ label }}</h3>
+        <div class="coupon__brand">
+            {% block coupon_teaser %}
+                {% if content.field_logo | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference.html.twig:19]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference.html.twig [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                    <div class="coupon__brand--logo">{{ content.field_logo }}</div>
+                {% endif %}
+            {% endblock %}
+        </div>
+        {% if content.field_coupon_discount | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference.html.twig:24]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference.html.twig [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                            <text id="_20" data-name="20" class="cls-2" style="fill: #fff; font-weight: 700;font-size: 32px; font-family: NotoSerif-Bold, Noto Serif;" transform="translate(1089.283 2598.393)">
+                                <tspan x="-24.231" y="0">{{ content.field_coupon_discount  | render | trim }}</tspan>
+                                <tspan style="font-size: 14px;" class="cls-3" y="0">%</tspan>
+                            </text>
+                        </g>
+                    </svg>
+                </span>
+            </div>
+        {% endif %}
+        {% if content.field_tnc | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference.html.twig]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference.html.twig [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+            <div class="coupon__tnc">
+                <div class="coupon__tnc-label expandable">Teilnahmebedingungen</div>
+                <div class="coupon__tnc-details">
+                    {{ content.field_tnc }}
+                </div>
+            </div>
+        {% endif %}
+        <div class="coupon__code--wrapper">
+            {% if content.field_coupon_code | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference.html.twig:52]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon--content-reference.html.twig [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+                <text id="Jetzt" style="fill: #fff;font-size: 14px;font-family: PTSans-NarrowBold, PT Sans Narrow;font-weight: 700;letter-spacing: 0.03em;" transform="translate(1098.317 2601.201)">
+                    <tspan x="-18.97" y="0">LEIDER</tspan>
+                    <tspan x="-27.377" y="15">VERPASST</tspan>
+                </text>
+            </g>
+        </svg>
+    </div>
+    {% if content.field_tracking_pixel | render is not empty %}
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon--content-reference.html.twig:76]: is there another way to check if the field is not empty?
Render is a pretty resource intensive operation

File: templates/node/coupon/node--coupon.html.twig [hunk 15]
```diff
@@ -76,7 +76,7 @@
 {% set classes = [
 'item-content',
```
Comment (mgoedecke) [inline templates/node/coupon/node--coupon.html.twig:77]: use indention

File: templates/node/newsletter/node--newsletter.html.twig [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+ * @todo Remove the id attribute (or make it a class), because if that gets
+ *   rendered twice on a page this is invalid CSS for example: two lists
+ *   in different view modes.
+ *
+ * @ingroup themeable
+ */
+#}
+
+{% set classes = [
+'region-full-content',
```
Comment (mgoedecke) [inline templates/node/newsletter/node--newsletter.html.twig]: use indention

File: js/behaviors/drupal.behaviors.coupons.js (hunk 1 above)
Comment (danschalow) [reply js/behaviors/drupal.behaviors.coupons.js:43]: We can, but `isExpired` is only called on the following querySelector:

```
document.querySelectorAll(".coupon[data-expires]")
```

Do you think it's still necessary?

**Pull request merged**